### PR TITLE
Campbell Moore DAE

### PR DESCRIPTION
**Pull request by elswit**
This is an index 3 DAE from this [paper](https://link.springer.com/content/pdf/10.1007/s002110050099.pdf). The index is reduced to 1 using matlab symbolic tools by introducing extra variables. In total 10 vars.  As a result, the mass and rhs files look messy. 
There is allegedly an initial condition that makes the trajectory of the solution periodic. I haven't been able to find it.  I'm working with my own consistent init condition.

### CODE REVIEW

File: src/+otp/+campbellmooredae/+presets/Canonical.m [hunk 1]
```diff
@@ -0,0 +1,15 @@
+classdef Canonical < otp.campbellmooredae.CampbellMooreDAEProblem
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/+presets/Canonical.m:1]: same thing as before with "*Problem"

File: src/+otp/+campbellmooredae/+presets/Canonical.m [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    methods
+        function obj = Canonical()
+            tspan = [0.0, 10*pi];
+            
+            
+            params.r = 10;
+            params.rho = 5;
+            
+            y0 = [15, 0, 0, 0, 0.039524144269346, -0.417592117550000, 0.042213548262602, -0.034980779193327, 0, 0]';
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/+presets/Canonical.m:10]: instead of ' use ;

File: src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    end
+    
+    methods (Access = protected)
+        function onSettingsChanged(obj)
+            r = obj.Parameters.r;
+            rho = obj.Parameters.rho;
+            
+            obj.Rhs = otp.Rhs(@(t, y) otp.campbellmooredae.f(t, y, r,rho), ...
+                otp.Rhs.FieldNames.Jacobian, @(t, y) otp.campbellmooredae.jac(t, y, r,rho), ...
+                otp.Rhs.FieldNames.Mass,  otp.campbellmooredae.mass([], [], r,rho), ...
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m:20]: "r, rho"

spacing

File: src/+otp/+campbellmooredae/+presets/Canonical.m (hunk 2 above)
Comment (AndreyAPopov) [reply src/+otp/+campbellmooredae/+presets/Canonical.m:10]: On the initial conditions: aren't they (or rather parts of them), given by equation 37 in the paper?

File: src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                otp.Rhs.FieldNames.MassSingular, 'yes');
+        end
+        
+        function validateNewState(obj, newTimeSpan, newY0, newParameters)
+            validateNewState@otp.Problem(obj, newTimeSpan, newY0, newParameters)
+            
+            otp.utils.StructParser(newParameters) ...
+                .checkField('rho', 'scalar', 'real', 'finite', 'positive');
+            otp.utils.StructParser(newParameters) ...
+                .checkField('r', 'scalar', 'real', 'finite', 'positive');
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m:31]: Reading the paper, I don't see a compelling reason for r and rho being positive, unless I am missing something major.

File: src/+otp/+campbellmooredae/jac.m [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
....*2.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,y(1,:).*1.0./sqrt(y(1,:).^2+y(2,:).^2).*(sqrt(y(1,:).^2+y(2,:).^2)-1.0e+1).*2.0,1.0./(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*(y(4,:).*(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*2.0-y(2,:).^2.*y(4,:).*2.0e+1+y(1,:).*y(2,:).*y(5,:).*2.0e+1).*2.0,0.0,0.0,-1.0,0.0,-1.0,0.0,0.0,0.0,y(2,:).*1.0./sqrt(y(1,:).^2+y(2,:).^2).*(sqrt(y(1,:).^2+y(2,:).^2)-1.0e+1).*2.0,1.0./(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*(y(5,:).*(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*2.0-y(1,:).^2.*y(5,:).*2.0e+1+y(1,:).*y(2,:).*y(4,:).*2.0e+1).*2.0,0.0,0.0,0.0,0.0,cos(t),sin(t),0.0,0.0,y(3,:).*2.0,y(6,:).*4.0,0.0,0.0,0.0,-1.0,y(1,:).*1.0./sqrt(y(1,:).^2+y(2,:).^2).*(sqrt(y(1,:).^2+y(2,:).^2)-1.0e+1).*2.0,y(2,:).*1.0./sqrt(y(1,:).^2+y(2,:).^2).*(sqrt(y(1,:).^2+y(2,:).^2)-1.0e+1).*2.0,y(3,:).*2.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,-1.0,0.0,0.0,0.0,0.0,y(1,:).*1.0./(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*(-(y(1,:).^2+y(2,:).^2).^(3.0./2.0)+y(1,:).^2.*1.0e+1+y(2,:).^2.*1.0e+1).*-2.0,1.0,0.0,0.0,0.0,0.0,-1.0,0.0,0.0,0.0,y(2,:).*1.0./(y(1,:).^2+y(2,:).^2).^(3.0./2.0).*(-(y(1,:).^2+y(2,:).^2).^(3.0./2.0)+y(1,:).^2.*1.0e+1+y(2,:).^2.*1.0e+1).*-2.0,0.0,0.0,0.0,0.0,0.0,0.0,-1.0,0.0,0.0,y(3,:).*2.0,0.0,1.0,0.0,0.0],[10,10]);
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/jac.m:3]: I can see that this was automatically generated, but it seems like it is in dire need of proper formatting.

File: src/+otp/+campbellmooredae/mass.m [hunk 6]
```diff
@@ -0,0 +1,5 @@
+function M = mass(~, ~, ~, ~)
+
+M = sparse([9,10,7,8],[1,3,4,6],[-1,-1,1,1],10,10);
```
Comment (AndreyAPopov) [inline src/+otp/+campbellmooredae/mass.m:3]: I know that this notation is very efficient, but again for this low amount of variables, are sparse matrices still worth it?

File: src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m (hunk 4 above)
Comment (AndreyAPopov) [reply src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m:31]: oh I see it now. In fact it says that r > rho.

we need to add, "@(x) x > newParameters.rho" to the validation for r I think.

Comment (AndreyAPopov) [reply src/+otp/+campbellmooredae/CampbellMooreDAEProblem.m:31]: another thing, we can more efficiently write this as
`otp.utils.StructParser(newParameters) ...
                .checkField('rho', 'scalar', 'real', 'finite', 'positive') ...
                .checkField('r', 'scalar', 'real', 'finite', 'positive', @(x) x > newParameters.rho);`

**Comment by Steven-Roberts** [comment]
Closing because it's become stale.

**Pull request closed without merging**